R/logistic_reg.R

```diff
@@ -0,0 +1,342 @@
+    processed = processed,
+    fit_intercept = fit_intercept,
+    penalty = penalty,
+    tol = tol,
+    C = C,
+    class_weight = class_weight,
+    sample_weight = sample_weight,
+    max_iters = max_iters,
+    linesearch_max_iters = linesearch_max_iters,
+    l1_ratio = l1_ratio
+  )
+}
+
+cuda_ml_logistic_reg_bridge <- function(processed,
+                                        fit_intercept,
+                                        penalty, tol, C,
+                                        class_weight, sample_weight,
+                                        max_iters, linesearch_max_iters,
+                                        l1_ratio,
+                                        ...) {
+  hardhat::validate_predictors_are_numeric(processed$predictors)
+  hardhat::validate_outcomes_are_univariate(processed$outcomes)
+  hardhat::validate_outcomes_are_factors(processed$outcomes)
+
+  x <- as.matrix(processed$predictors)
+  # convert outcomes to 0-based enum values
+  y <- as.integer(processed$outcomes[[1]]) - 1L
+
+  n_classes <- nlevels(processed$outcomes[[1]])
+  loss_type <- ifelse(
+    n_classes > 2,
+    logistic_reg_loss_type$softmax,
+    logistic_reg_loss_type$sigmoid
+  )
+
+  qn_params <- logistic_reg_build_qn_params(C, l1_ratio, penalty)
+
+  if (!is.null(class_weight)) {
+    logistic_reg_validate_class_weight(class_weight, processed)
+  }
+
+  if (!is.null(sample_weight)) {
+    logistic_reg_validate_sample_weight(sample_weight, processed)
+  } else {
+    sample_weight <- rep(1, nrow(x))
+  }
+  sample_weight <- logistic_reg_build_sample_weight(
+    sample_weight, class_weight, processed
+  )
+
+  model_xptr <- .qn_fit(
+    X = x, y = y,
+    n_classes = n_classes,
+    loss_type = loss_type,
+    fit_intercept = fit_intercept,
+    l1 = qn_params$l1, l2 = qn_params$l2,
+    max_iters = as.integer(max_iters), tol = as.numeric(tol),
+    delta = as.numeric(tol) * 0.01,
+    linesearch_max_iters = as.integer(linesearch_max_iters),
+    lbfgs_memory = 5L,
+    sample_weight = as.numeric(sample_weight)
+  )
+
+  new_model(
+    cls = "cuda_ml_logistic_reg",
+    xptr = model_xptr,
+    blueprint = processed$blueprint
+  )
+}
+
+#' Make predictions on new data points.
+#'
+#' Make predictions on new data points using a CuML logistic regression model.
+#'
+#' @template predict
+#'
+#' @importFrom ellipsis check_dots_used
+#' @export
+predict.cuda_ml_logistic_reg <- function(object, x, ...) {
+  check_dots_used()
+
+  processed <- hardhat::forge(x, object$blueprint)
+  model <- object$xptr
+
+  preds <- .qn_predict(
+    X = as.matrix(processed$predictors),
+    n_classes = model$n_classes,
+    coefs = model$coefs,
+    loss_type = model$loss_type,
+    fit_intercept = model$fit_intercept
+  )
+  # convert from 0-based internal representation to 1-based representation which
+  # will correspond to the factor levels of all possible outcomes in R
+  preds <- preds + 1L
+
+  postprocess_classification_results(preds, object)
```

Comment:
I followed what's happening here (can't link to it within a GH review for some reason)
calling `factor()` is kind of unnecessarily expensive, since it call's `as.character()` on `preds`, even though that's not necessary here, since you already have integers. 
It might be better to just building the factor object directly:
```
attributes(preds) <- list(class = "factor", levels = <>)
```
or something similar.